### PR TITLE
Query visit dimension

### PR DESCRIPTION
**Pull request by gijskant**
- Enable querying on visit dimension fields (e.g., on visit start date).
- Reorder parameters of highdim call in QueryService (to align with other calls).

### CODE REVIEW

File: transmart-core-db/src/main/groovy/org/transmartproject/db/multidimquery/query/DimensionMetadata.groovy [hunk 1]
```diff
@@ -101,7 +101,11 @@ class DimensionMetadata {
 
     static final List<Field> getSupportedFields() {
         dimensionMetadataMap.values().collectMany {
-            (it.type in [DimensionFetchType.COLUMN, DimensionFetchType.TABLE]) ? it.fields : [] as List<Field> }
+            (it.type in [
+                    DimensionFetchType.COLUMN,
+                    DimensionFetchType.TABLE,
+                    DimensionFetchType.VISIT
+            ]) ? it.fields : [] as List<Field> }
```
Comment (JanKanis) [inline transmart-core-db/src/main/groovy/org/transmartproject/db/multidimquery/query/DimensionMetadata.groovy]: The `as List<Field>` cast seems superfluous, remove it?

File: transmart-core-db/src/main/groovy/org/transmartproject/db/multidimquery/query/HibernateCriteriaQueryBuilder.groovy [hunk 2]
```diff
@@ -237,101 +240,149 @@ class HibernateCriteriaQueryBuilder implements QueryBuilder<Criterion, DetachedC
...
-        constraint.value = convertValue(constraint.field, constraint.value)
-        String propertyName = getFieldPropertyName(constraint.field)
-        switch(constraint.operator) {
+    static Criterion applyOperator(Operator operator, String propertyName, Type type, Object value) {
+        Criterion criterion
+        switch(operator) {
             case Operator.EQUALS:
-                return Restrictions.eq(propertyName, constraint.value)
+                criterion = Restrictions.eq(propertyName, value)
+                break
```
Comment (JanKanis) [inline transmart-core-db/src/main/groovy/org/transmartproject/db/multidimquery/query/HibernateCriteriaQueryBuilder.groovy]: This looks like a significant uglification with all the `break`s now needed. How about wrapping the switch in its own method so you can keep using return?

Comment (JanKanis) [inline transmart-core-db/src/main/groovy/org/transmartproject/db/multidimquery/query/HibernateCriteriaQueryBuilder.groovy]: Or even make this into a static Map

    import static Operator.*
    [(EQUALS): {propName, value -> Restrictions.eq(propName, value) },
     (NOT_EQUALS): {....

I'm not sure if that is a net improvement in readability, so just for consideration.

File: transmart-core-db/src/main/groovy/org/transmartproject/db/multidimquery/query/HibernateCriteriaQueryBuilder.groovy [hunk 3]
```diff
@@ -237,101 +240,149 @@ class HibernateCriteriaQueryBuilder implements QueryBuilder<Criterion, DetachedC
...
+        constraint.value = convertValue(constraint.field, constraint.value)
+        if (constraint.field.dimension == VisitDimension) {
+            /**
+             * special case that requires a subquery, because there is no proper
+             * reference to the visit dimension in {@link ObservationFact}.
+             */
+            DetachedCriteria subCriteria = DetachedCriteria.forClass(org.transmartproject.db.i2b2data.VisitDimension, 'visit')
+            String propertyName = constraint.field.fieldName
+            subCriteria.add(applyOperator(constraint.operator, propertyName, constraint.field.type, constraint.value))
+            return Subqueries.propertiesIn(['encounterNum', 'patient'] as String[],
```
Comment (JanKanis) [inline transmart-core-db/src/main/groovy/org/transmartproject/db/multidimquery/query/HibernateCriteriaQueryBuilder.groovy]: Doesn't groovy do this `String[]` conversion automatically to match the value to the parameter type?

File: transmart-core-db/src/main/groovy/org/transmartproject/db/multidimquery/query/DimensionMetadata.groovy (hunk 1 above)
Comment (gijskant) [inline transmart-core-db/src/main/groovy/org/transmartproject/db/multidimquery/query/DimensionMetadata.groovy]: @JanKanis Apparently not:
```
/home/travis/build/thehyve/transmart-upgrade/transmart-core-db/src/main/groovy/org/transmartproject/db/multidimquery/query/DimensionMetadata.groovy: 103: [Static type checking] - Incompatible generic argument types. Cannot assign java.util.List <?> to: java.util.List <Field>

 @ line 103, column 9.

           dimensionMetadataMap.values().collectMany {
```

Comment (gijskant) [inline transmart-core-db/src/main/groovy/org/transmartproject/db/multidimquery/query/DimensionMetadata.groovy]: I will revert the latest commit.

**Comment by JanKanis** [comment]
Included in #92

**Pull request merged**